overlord/snapstate/handlers.go

```diff
@@ -1042,20 +1090,40 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 	// record type
 	snapst.SetType(newInfo.GetType())
 
+	pb := NewTaskProgressAdapterLocked(t)
+
 	// XXX: this block is slightly ugly, find a pattern when we have more examples
 	model, _ := ModelFromTask(t)
 	err = m.backend.LinkSnap(newInfo, model, perfTimings)
-	if err != nil {
-		pb := NewTaskProgressAdapterLocked(t)
-		err := m.backend.UnlinkSnap(newInfo, pb)
-		if err != nil {
-			t.Errorf("cannot cleanup failed attempt at making snap %q available to the system: %v", snapsup.InstanceName(), err)
+	// defer a cleanup helper which will unlink the snap if anything fails after
+	// this point
+	defer func() {
+		if err == nil {
+			return
 		}
+		// err is not nil, we need to try and unlink the snap to cleanup after
+		// ourselves
+		unlinkErr := m.backend.UnlinkSnap(newInfo, pb)
+		if unlinkErr != nil {
+			t.Errorf("cannot cleanup failed attempt at making snap %q available to the system: %v", snapsup.InstanceName(), unlinkErr)
+		}
+	}()
+	if err != nil {
+		return err
 	}
+
+	// Restore disabled services
+	var missingSvcs []string
+	missingSvcs, err = m.backend.RestoreDisabledServices(newInfo, snapst.LastActiveDisabledServices, pb)
```

Comment:
The same comment about locking the state whiel running external commands. You may need to use `NewTaskProgressAdapterUnlocked()` too

Reply:
Yep, applied this here to